enhancements/cgu-precaching-additional-images.md

```diff
@@ -0,0 +1,308 @@
+as the CGU on the hub cluster.
+   - Example CGU CR:
+    ```yaml
+    apiVersion: ran.openshift.io/v1alpha1
+    kind: ClusterGroupUpgrade
+    metadata:
+      name: cgu
+      namespace: default
+    spec:
+      clusters:
+      - sno1
+      - sno3
+      - sno7
+      enable: false
+      preCaching: true
+      backup: false
+      managedPolicies:
+      - policy0-common-config-policy
+      remediationStrategy:
+        maxConcurrency: 4
+        timeout: 240
+    ```
+   - Example ConfigMap:
+    ```yaml
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: cluster-group-upgrade-overrides
+      namespace: default
+    data:
+      additionalWorkloadImages: |
+        quay.io/foobar/application1@sha256:3d5800990dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47e2e1ef
+        quay.io/foobar/application2@sha256:3d5800123dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47adfaef
+        ...
+        quay.io/foobar/applicationN@sha256:4fe1334adfafadsf987123adfffdaf1243340adfafdedga0991234afdadfsa09
+    ```
+
+2. The CR is validated (a function verifies if the clusters specified in the CR are part of a `ManagedCluster`
+object).
+
+3. If the validation succeeds, the pre-caching reconciliation commences. Note that all the sites are
+pre-cached concurrently.
+
+4. The pre-cache `spec` is then validated by checking if the managed policies exist. A valid `spec` will
+result in the following condition:
+    ```text
+    Type:    "PrecacheSpecValid",
+    Status:  True,
+    Reason:  "PrecacheSpecIsWellFormed",
+    Message: "Precaching spec is valid and consistent"
+    ```
+
+5. Pre-caching commences upon successful validation of the pre-cache `spec`. The following conditions are
+set while pre-caching is in-progress:
+    ```text
+    Type:    "PrecachingSuceeded",
+    Status:  False,
+    Reason:  "InProgress",
+    Message: "Precaching is required and not done"
+    ```
+
+6. After the pre-caching is successfully completed, the following condition is set:
+    ```text
+    Type:    "PrecachingSuceeded",
+    Status:  True,
+    Reason:  "PrecachingCompleted",
+    Message: "Precaching is completed for all clusters"
+    ```
+
+7. The remaining workflow for TALM continues as normal from here onwards.
+
+#### Variation [optional]
+
+N/A
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+
+N/A
+
+
+### Risks and Mitigations
+
+N/A
+
+### Benefits
+
+The benefits of the proposed approach are as follows:
+- The solution extends the current pre-caching TALM functionality in a unified approach.
+- The proposal integrates with TALM to control the rollout of policy (ConfigMap) to the managed
+clusters.
+- No additional operators need to be installed (apart from TALM) and thus compute requirements
+remain unchanged.
+
+### Drawbacks
+
+A drawback of the proposed approach is that TALM now accepts the additional workload of pre-caching
```

Comment:
Is it really a drawback? I see that as a further enhancement

Reply:
The reason why this is listed as a drawback is because there is additional responsibility for TALM which implies a potential minor increase in resource utilization. This is to contrast the limitation of the alternative solution which entails isolating the pre-caching application and developing it as its own operator. In the case of the alternative solution, TALM would have reduced computational footprint.

Reply:
That is correct Ian - the pre-caching job is indeed executed on the spoke. There might be an insignificant added cost of parsing the `additionalImages` field on the hub cluster. This is not an expensive computational operation - so should be okay! Just wanted to highlight this here.

Reply:
That is correct Ian - the pre-caching job is indeed executed on the spoke. There might be an insignificant added cost of parsing the `additionalImages` field on the hub cluster. This is not an expensive computational operation - so should be okay! Just wanted to highlight this here.

Reply:
The drawback has been updated to highlight the above comments.